File: src/main/java/net/insprill/cjm/messages/types/ChatMessage.java
package net.insprill.cjm.messages.types;

import net.insprill.cjm.message.MessageVisibility;
import net.insprill.cjm.utils.ChatUtils;
import net.insprill.xenlib.CenteredMessages;
import net.insprill.xenlib.files.YamlFile;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;

import java.io.File;
import java.util.List;

public class ChatMessage implements MessageType {

    private static final String CENTER_PREFIX = "center:";

    private final YamlFile config = new YamlFile("messages" + File.separator + "chat.yml", false);

    @Override
    public YamlFile getConfig() {
        return config;
    }

    @Override
    public String getKey() {
        return "Messages";
    }

    @Override
    public String getName() {
        return "chat";
    }

    @Override
    public void handle(Player primaryPlayer, List<Player> players, String rootPath, String chosenPath, MessageVisibility visibility) {
        List<String> messages = config.getStringList(chosenPath);
        ChatUtils.setPlaceholders(primaryPlayer, messages);
        messages.replaceAll(s -> (s.startsWith(CENTER_PREFIX))
                ? CenteredMessages.centerMessage(s.substring(CENTER_PREFIX.length()))
                : s);
        for (String msg : messages) {
            boolean isJson = msg.contains("{\"text\":");
            for (Player player : players) {
                if (isJson) {
                    Bukkit.dispatchCommand(Bukkit.getConsoleSender(), "minecraft:tellraw " + player.getName() + " " + msg);
                } else {
                    player.sendMessage(msg);
                }
            }
            if (visibility == MessageVisibility.PUBLIC && !isJson) {
                Bukkit.getConsoleSender().sendMessage(msg);
            }
        }
    }

}

File: src/main/kotlin/net/insprill/cjm/message/types/ChatMessage.kt
package net.insprill.cjm.message.types

import de.themoep.minedown.MineDown
import net.insprill.cjm.message.MessageVisibility
import net.insprill.cjm.placeholder.Placeholders.Companion.fillPlaceholders
import net.insprill.xenlib.CenteredMessages
import net.insprill.xenlib.files.YamlFile
import net.md_5.bungee.api.ChatMessageType
import org.bukkit.entity.Player
import java.io.File

class ChatMessage : MessageType {

    private val CENTER_PREFIX = "center:"

    override val config = YamlFile("messages" + File.separator + "chat.yml", false)
    override val key = "Messages"
    override val name = "chat"

    override fun handle(primaryPlayer: Player, players: List<Player>, rootPath: String?, chosenPath: String, visibility: MessageVisibility) {
        val messages = config.getStringList(chosenPath)
        formatMessages(primaryPlayer, messages)
        val components = messages.flatMap { MineDown.parse(it).toList() }.toTypedArray()
        for (player in players) {
            @Suppress("DEPRECATION")
            player.spigot().sendMessage(ChatMessageType.CHAT, *components)
        }
    }

    private fun formatMessages(primaryPlayer: Player, messages: MutableList<String>) {
        fillPlaceholders(primaryPlayer, messages)
        messages.replaceAll {
            if (it.startsWith(CENTER_PREFIX)) {
                CenteredMessages.centerMessage(it.substring(CENTER_PREFIX.length))
            } else {
                it
            }
        }
    }

}
